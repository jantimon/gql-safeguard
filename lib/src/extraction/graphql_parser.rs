//! GraphQL Parser and Directive Extraction
//!
//! This module provides functionality for parsing GraphQL strings into structured AST
//! representations and extracting GraphQL directives for analysis. It handles the
//! conversion from raw GraphQL template literals (extracted by `typescript_parser`)
//! into typed data structures suitable for directive validation and dependency analysis.
//!
//! # Architecture
//!
//! The parsing pipeline consists of two main phases:
//!
//! ## 1. GraphQL AST Parsing
//! - **Input**: Raw GraphQL strings from TypeScript template literals
//! - **Process**: Parse using `graphql-parser` crate for syntax validation
//! - **Output**: Structured `GraphQLItem` representations (queries and fragments)
//!
//! ## 2. Directive Extraction
//! - **Input**: Parsed GraphQL AST nodes
//! - **Process**: Extract `@catch` and `@throwOnFieldError` directives with locations
//! - **Output**: Typed directive data structures for validation analysis
//!
//! # Supported GraphQL Constructs
//!
//! - **Queries**: Named and anonymous query operations with field selections
//! - **Fragments**: Named fragment definitions with type conditions
//! - **Fragment Spreads**: References to other fragments via `...FragmentName`
//! - **Directives**: `@catch` and `@throwOnFieldError` directives on various locations
//! - **Fields**: Individual field selections with optional directives

use std::path::PathBuf;

use crate::extraction::typescript_parser::GraphQLString;
use anyhow::Result;
use graphql_parser::parse_query;
use graphql_parser::query::{
    Definition, Document as QueryDocument, OperationDefinition, Selection, SelectionSet,
};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum DirectiveType {
    Catch,
    ThrowOnFieldError,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Directive {
    pub directive_type: DirectiveType,
    pub position: u32,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum GraphQLItem {
    Query(QueryOperation),
    Fragment(FragmentDefinition),
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct QueryOperation {
    pub name: String,
    pub fields: Vec<Field>,
    pub fragments: Vec<FragmentSpread>,
    pub directives: Vec<Directive>,
    pub file_path: PathBuf,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct FragmentDefinition {
    pub name: String,
    pub fields: Vec<Field>,
    pub fragments: Vec<FragmentSpread>,
    pub directives: Vec<Directive>,
    pub file_path: PathBuf,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Field {
    pub name: String,
    pub directives: Vec<Directive>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct FragmentSpread {
    pub name: String,
    pub directives: Vec<Directive>,
}

/// Parses a GraphQL string into structured AST items with directive extraction.
///
/// This function takes a raw GraphQL string (extracted from TypeScript template literals)
/// and converts it into typed data structures representing queries, fragments, and their
/// associated directives. It performs syntax validation and extracts only the information
/// relevant for directive protection analysis
pub fn parse_graphql_to_ast(graphql_string: &GraphQLString) -> Result<Vec<GraphQLItem>> {
    // Parse the GraphQL string using the graphql-parser library
    let document: QueryDocument<String> = parse_query(&graphql_string.content).map_err(|e| {
        anyhow::anyhow!(
            "GraphQL syntax error in {} at position {}: {:?}",
            graphql_string.file_path.display(),
            graphql_string.position,
            e
        )
    })?;

    let mut items = Vec::new();

    // Process each definition in the GraphQL document
    for definition in document.definitions {
        match definition {
            Definition::Operation(op) => {
                if let Some(query) = convert_operation_to_query(
                    op,
                    &graphql_string.file_path,
                    graphql_string.position,
                )? {
                    items.push(GraphQLItem::Query(query));
                }
            }
            Definition::Fragment(frag) => {
                let fragment = convert_fragment_definition(
                    frag,
                    &graphql_string.file_path,
                    graphql_string.position,
                )?;
                items.push(GraphQLItem::Fragment(fragment));
            }
        }
    }

    Ok(items)
}

/// Converts a GraphQL operation definition to our internal QueryOperation representation.
///
/// This function processes query operations from the parsed GraphQL AST and extracts
/// the information needed for directive validation. It focuses on query operations
/// since mutations and subscriptions are not typically relevant for `@throwOnFieldError`
/// protection analysis
fn convert_operation_to_query(
    op: OperationDefinition<String>,
    file_path: &std::path::Path,
    position: u32,
) -> Result<Option<QueryOperation>> {
    match op {
        OperationDefinition::Query(query) => {
            // Assign a name to anonymous queries for tracking purposes
            let name = query.name.unwrap_or_else(|| "AnonymousQuery".to_string());

            // Extract directives from the query operation
            let directives = extract_directives_from_directive_list(&query.directives, position);

            // Process the selection set to extract fields and fragment spreads
            let (fields, fragments) =
                convert_selection_set_to_fields_and_spreads(&query.selection_set, position);

            Ok(Some(QueryOperation {
                name,
                fields,
                fragments,
                directives,
                file_path: file_path.to_path_buf(),
            }))
        }
        OperationDefinition::Mutation(_) | OperationDefinition::Subscription(_) => {
            // Skip mutations and subscriptions as they're not typically relevant
            // for @throwOnFieldError protection analysis
            Ok(None)
        }
        OperationDefinition::SelectionSet(_) => {
            // Skip bare selection sets for now - these are less common in practice
            // and would need additional handling for anonymous operation tracking
            Ok(None)
        }
    }
}

/// Converts a GraphQL fragment definition to our internal FragmentDefinition representation.
///
/// This function processes fragment definitions from the parsed GraphQL AST and extracts
/// all information needed for directive validation and dependency analysis. Fragments
/// are critical for protection analysis since `@catch` directives on fragments can
/// protect nested `@throwOnFieldError` directives
fn convert_fragment_definition(
    frag: graphql_parser::query::FragmentDefinition<String>,
    file_path: &std::path::Path,
    position: u32,
) -> Result<FragmentDefinition> {
    // Extract directives from the fragment definition
    let directives = extract_directives_from_directive_list(&frag.directives, position);

    // Process the selection set to extract fields and fragment spreads
    let (fields, fragments) =
        convert_selection_set_to_fields_and_spreads(&frag.selection_set, position);

    Ok(FragmentDefinition {
        name: frag.name,
        fields,
        fragments,
        directives,
        file_path: file_path.to_path_buf(),
    })
}

/// Converts a GraphQL selection set into separate lists of fields and fragment spreads.
///
/// This function processes the selection set from queries or fragments and separates
/// the different types of selections into appropriate data structures. This separation
/// is important for directive validation since protection rules apply differently to
/// direct field selections versus fragment spreads
fn convert_selection_set_to_fields_and_spreads(
    selection_set: &SelectionSet<String>,
    position: u32,
) -> (Vec<Field>, Vec<FragmentSpread>) {
    let mut fields = Vec::new();
    let mut fragments = Vec::new();

    for selection in &selection_set.items {
        match selection {
            Selection::Field(field) => {
                // Extract directives from the field selection
                let directives =
                    extract_directives_from_directive_list(&field.directives, position);
                fields.push(Field {
                    name: field.name.clone(),
                    directives,
                });

                // Recursively process nested selection set if it has items
                if !field.selection_set.items.is_empty() {
                    let (_nested_fields, nested_fragments) =
                        convert_selection_set_to_fields_and_spreads(&field.selection_set, position);
                    fragments.extend(nested_fragments);
                }
            }
            Selection::FragmentSpread(spread) => {
                // Extract directives from the fragment spread
                let directives =
                    extract_directives_from_directive_list(&spread.directives, position);
                fragments.push(FragmentSpread {
                    name: spread.fragment_name.clone(),
                    directives,
                });
            }
            Selection::InlineFragment(_) => {
                // Skip inline fragments for now - they're less common and would
                // require additional complexity for proper directive inheritance
            }
        }
    }

    (fields, fragments)
}

/// Extracts relevant GraphQL directives from a list of directive AST nodes.
///
/// This function processes the directive list from the GraphQL parser and converts
/// only the directives relevant to protection analysis (`@catch` and `@throwOnFieldError`)
/// into our internal `Directive` representation. Other directives are ignored since
/// they don't affect the safety analysis
fn extract_directives_from_directive_list(
    directives: &[graphql_parser::query::Directive<String>],
    position: u32,
) -> Vec<Directive> {
    directives
        .iter()
        .filter_map(|dir| {
            // Only process directives relevant to protection analysis
            let directive_type = match dir.name.as_str() {
                "catch" => DirectiveType::Catch,
                "throwOnFieldError" => DirectiveType::ThrowOnFieldError,
                _ => return None,
            };

            Some(Directive {
                directive_type,
                position,
            })
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::extraction::typescript_parser;
    use std::fs;
    use std::path::PathBuf;

    /// Formats GraphQL AST items for snapshot testing with detailed structure visualization
    fn format_graphql_ast_result(
        file_path: &std::path::Path,
        graphql_items: &[GraphQLItem],
    ) -> String {
        // Convert to relative path from git root for portable snapshots
        let git_root = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .parent()
            .unwrap()
            .to_path_buf();
        let relative_path = file_path.strip_prefix(&git_root).unwrap_or(file_path);

        let mut result = format!("File: {}\n", relative_path.display());
        result.push_str(&format!("GraphQL AST items: {}\n\n", graphql_items.len()));

        for (i, item) in graphql_items.iter().enumerate() {
            result.push_str(&format!("=== AST Item {} ===\n", i + 1));

            match item {
                GraphQLItem::Query(query) => {
                    result.push_str(&format!("Type: Query\n"));
                    result.push_str(&format!("Name: {}\n", query.name));

                    // Use relative path for file location in AST items too
                    let query_file_relative = query
                        .file_path
                        .strip_prefix(&git_root)
                        .unwrap_or(&query.file_path);
                    result.push_str(&format!("File: {}\n", query_file_relative.display()));

                    // Show query-level directives with emojis
                    result.push_str(&format!("Directives: {}\n", query.directives.len()));
                    for directive in &query.directives {
                        let emoji = match directive.directive_type {
                            DirectiveType::Catch => "🧤",
                            DirectiveType::ThrowOnFieldError => "☄️",
                        };
                        result.push_str(&format!("  - {:?} {}\n", directive.directive_type, emoji));
                    }

                    // Show fields with their directives
                    result.push_str(&format!("Fields: {}\n", query.fields.len()));
                    for field in &query.fields {
                        result.push_str(&format!("  - {}", field.name));
                        if !field.directives.is_empty() {
                            result.push_str(" [");
                            for (j, directive) in field.directives.iter().enumerate() {
                                if j > 0 {
                                    result.push_str(", ");
                                }
                                let emoji = match directive.directive_type {
                                    DirectiveType::Catch => "🧤",
                                    DirectiveType::ThrowOnFieldError => "☄️",
                                };
                                result
                                    .push_str(&format!("{:?} {}", directive.directive_type, emoji));
                            }
                            result.push_str("]");
                        }
                        result.push_str("\n");
                    }

                    // Show fragment spreads
                    result.push_str(&format!("Fragment Spreads: {}\n", query.fragments.len()));
                    for fragment in &query.fragments {
                        result.push_str(&format!("  - {}", fragment.name));
                        if !fragment.directives.is_empty() {
                            result.push_str(" [");
                            for (j, directive) in fragment.directives.iter().enumerate() {
                                if j > 0 {
                                    result.push_str(", ");
                                }
                                let emoji = match directive.directive_type {
                                    DirectiveType::Catch => "🧤",
                                    DirectiveType::ThrowOnFieldError => "☄️",
                                };
                                result
                                    .push_str(&format!("{:?} {}", directive.directive_type, emoji));
                            }
                            result.push_str("]");
                        }
                        result.push_str("\n");
                    }
                }
                GraphQLItem::Fragment(fragment) => {
                    result.push_str(&format!("Type: Fragment\n"));
                    result.push_str(&format!("Name: {}\n", fragment.name));

                    // Use relative path for file location in AST items too
                    let fragment_file_relative = fragment
                        .file_path
                        .strip_prefix(&git_root)
                        .unwrap_or(&fragment.file_path);
                    result.push_str(&format!("File: {}\n", fragment_file_relative.display()));

                    // Show fragment-level directives with emojis
                    result.push_str(&format!("Directives: {}\n", fragment.directives.len()));
                    for directive in &fragment.directives {
                        let emoji = match directive.directive_type {
                            DirectiveType::Catch => "🧤",
                            DirectiveType::ThrowOnFieldError => "☄️",
                        };
                        result.push_str(&format!("  - {:?} {}\n", directive.directive_type, emoji));
                    }

                    // Show fields with their directives
                    result.push_str(&format!("Fields: {}\n", fragment.fields.len()));
                    for field in &fragment.fields {
                        result.push_str(&format!("  - {}", field.name));
                        if !field.directives.is_empty() {
                            result.push_str(" [");
                            for (j, directive) in field.directives.iter().enumerate() {
                                if j > 0 {
                                    result.push_str(", ");
                                }
                                let emoji = match directive.directive_type {
                                    DirectiveType::Catch => "🧤",
                                    DirectiveType::ThrowOnFieldError => "☄️",
                                };
                                result
                                    .push_str(&format!("{:?} {}", directive.directive_type, emoji));
                            }
                            result.push_str("]");
                        }
                        result.push_str("\n");
                    }

                    // Show fragment spreads
                    result.push_str(&format!("Fragment Spreads: {}\n", fragment.fragments.len()));
                    for spread in &fragment.fragments {
                        result.push_str(&format!("  - {}", spread.name));
                        if !spread.directives.is_empty() {
                            result.push_str(" [");
                            for (j, directive) in spread.directives.iter().enumerate() {
                                if j > 0 {
                                    result.push_str(", ");
                                }
                                let emoji = match directive.directive_type {
                                    DirectiveType::Catch => "🧤",
                                    DirectiveType::ThrowOnFieldError => "☄️",
                                };
                                result
                                    .push_str(&format!("{:?} {}", directive.directive_type, emoji));
                            }
                            result.push_str("]");
                        }
                        result.push_str("\n");
                    }
                }
            }
            result.push_str("\n");
        }

        result
    }

    /// Processes all GraphQL strings from a fixture directory and parses them to AST
    fn process_fixture_directory_to_ast(dir_name: &str) -> String {
        let fixture_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .parent()
            .unwrap()
            .join("fixtures")
            .join(dir_name);

        let mut results = Vec::new();

        if let Ok(entries) = fs::read_dir(&fixture_dir) {
            let mut files: Vec<_> = entries
                .filter_map(|entry| entry.ok())
                .filter(|entry| {
                    let path = entry.path();
                    path.is_file()
                        && (path.extension() == Some(std::ffi::OsStr::new("ts"))
                            || path.extension() == Some(std::ffi::OsStr::new("tsx")))
                })
                .collect();

            // Sort files by name for consistent snapshot ordering
            files.sort_by_key(|entry| entry.file_name());

            for entry in files {
                let file_path = entry.path();

                // First extract GraphQL strings from TypeScript
                match typescript_parser::extract_graphql_from_file(&file_path) {
                    Ok(graphql_strings) => {
                        // Then parse each GraphQL string to AST
                        for graphql_string in graphql_strings {
                            match parse_graphql_to_ast(&graphql_string) {
                                Ok(graphql_items) => {
                                    let result =
                                        format_graphql_ast_result(&file_path, &graphql_items);
                                    results.push(result);
                                }
                                Err(e) => {
                                    results.push(format!(
                                        "File: {}\nGraphQL Parse Error: {}\nContent: {}\n\n",
                                        file_path.display(),
                                        e,
                                        graphql_string.content
                                    ));
                                }
                            }
                        }
                    }
                    Err(e) => {
                        results.push(format!(
                            "File: {}\nTypeScript Parse Error: {}\n\n",
                            file_path.display(),
                            e
                        ));
                    }
                }
            }
        }

        results.join("---\n\n")
    }

    /// Tests GraphQL AST parsing from all valid fixture files
    #[test]
    fn test_parse_valid_fixtures_to_ast() {
        let result = process_fixture_directory_to_ast("valid");
        insta::assert_snapshot!(result);
    }

    /// Tests GraphQL AST parsing from all invalid fixture files
    #[test]
    fn test_parse_invalid_fixtures_to_ast() {
        let result = process_fixture_directory_to_ast("invalid");
        insta::assert_snapshot!(result);
    }

    /// Tests GraphQL AST parsing from all edge case fixture files
    #[test]
    fn test_parse_edge_case_fixtures_to_ast() {
        let result = process_fixture_directory_to_ast("edge_cases");
        insta::assert_snapshot!(result);
    }
}
