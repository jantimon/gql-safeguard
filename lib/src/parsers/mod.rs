//! Two-stage parsing: extract GraphQL from TypeScript, then parse to AST
//!
//! Separation enables robust extraction from complex TS/TSX without GraphQL syntax errors.

pub mod graphql_parser;
pub mod typescript_parser;
