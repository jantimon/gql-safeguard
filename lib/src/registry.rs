//! GraphQL Extraction Registry
//!

/// A dashmap registry with:
/// process_files()
///
/// a dashmap regisry for all fragments (key fragment name)
/// a dashmap registry for all queries (key query name)
use anyhow::Result;
use dashmap::DashMap;
use globset::{Glob, GlobSetBuilder};
use ignore::{WalkBuilder, WalkState};
use rayon::prelude::*;
use serde::{Deserialize, Serialize};
use std::path::Path;
use std::sync::Arc;

use crate::parsers::graphql_parser::{
    parse_graphql_to_ast, FragmentDefinition, GraphQLItem, QueryOperation,
};
use crate::parsers::typescript_parser::extract_graphql_from_file;

/// Registry for GraphQL fragments
pub type FragmentRegistry = Arc<DashMap<String, FragmentDefinition>>;

/// Registry for GraphQL queries
pub type QueryRegistry = Arc<DashMap<String, QueryOperation>>;

/// Main registry that holds both fragments and queries
#[derive(Serialize, Deserialize)]
pub struct GraphQLRegistry {
    #[serde(with = "serde_dashmap")]
    pub fragments: FragmentRegistry,
    #[serde(with = "serde_dashmap")]
    pub queries: QueryRegistry,
}

/// Custom serialization for DashMap
mod serde_dashmap {
    use super::*;
    use serde::{Deserializer, Serializer};
    use std::collections::HashMap;

    pub fn serialize<S, K, V>(
        dashmap: &Arc<DashMap<K, V>>,
        serializer: S,
    ) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
        K: Clone + Serialize + Eq + std::hash::Hash,
        V: Clone + Serialize,
    {
        let map: HashMap<K, V> = dashmap
            .iter()
            .map(|entry| (entry.key().clone(), entry.value().clone()))
            .collect();
        map.serialize(serializer)
    }

    pub fn deserialize<'de, D, K, V>(deserializer: D) -> Result<Arc<DashMap<K, V>>, D::Error>
    where
        D: Deserializer<'de>,
        K: Clone + Deserialize<'de> + Eq + std::hash::Hash,
        V: Clone + Deserialize<'de>,
    {
        let map: HashMap<K, V> = HashMap::deserialize(deserializer)?;
        let dashmap = DashMap::new();
        for (k, v) in map {
            dashmap.insert(k, v);
        }
        Ok(Arc::new(dashmap))
    }
}

impl Default for GraphQLRegistry {
    fn default() -> Self {
        Self::new()
    }
}

impl GraphQLRegistry {
    pub fn new() -> Self {
        Self {
            fragments: Arc::new(DashMap::new()),
            queries: Arc::new(DashMap::new()),
        }
    }
}

/// Function which returns a new GraphQL registry for the given file list
pub fn process_files(files: &[String]) -> GraphQLRegistry {
    let registry = GraphQLRegistry::new();

    files.par_iter().for_each(|file| {
        parse_file(Path::new(file), &registry);
    });

    registry
}

/// Function which processes files matching a glob pattern using streaming
pub fn process_glob(
    root_path: &Path,
    include_patterns: &[&str], // e.g. &["**/*.ts", "**/*.tsx"]
    exclude_patterns: &[&str], // e.g. &["**/node_modules/**"]
) -> Result<GraphQLRegistry> {
    // Build include GlobSet
    let mut include_builder = GlobSetBuilder::new();
    for pattern in include_patterns {
        include_builder.add(Glob::new(pattern)?);
    }
    let include_set = Arc::new(include_builder.build()?);

    // Build exclude GlobSet
    let mut exclude_builder = GlobSetBuilder::new();
    for pattern in exclude_patterns {
        exclude_builder.add(Glob::new(pattern)?);
    }
    let exclude_set = Arc::new(exclude_builder.build()?);

    let registry = GraphQLRegistry::new();
    let registry_ref = &registry;

    WalkBuilder::new(root_path)
        .standard_filters(false)
        .build_parallel()
        .run(|| {
            let include = Arc::clone(&include_set);
            let exclude = Arc::clone(&exclude_set);
            let registry = registry_ref;

            Box::new(move |entry_res: Result<ignore::DirEntry, ignore::Error>| {
                if let Ok(entry) = entry_res {
                    let path = entry.path();
                    if path.is_dir() && exclude.is_match(path) {
                        return WalkState::Skip;
                    } else if path.is_file() && include.is_match(path) {
                        parse_file(path, registry);
                    }
                }
                WalkState::Continue
            })
        });

    Ok(registry)
}

fn parse_file(file: &Path, registry: &GraphQLRegistry) {
    let graphql_strings_result = extract_graphql_from_file(Path::new(file));
    if let Ok(graphql_strings) = graphql_strings_result {
        for graphql_string in &graphql_strings {
            let graphql_ast = parse_graphql_to_ast(graphql_string);
            if let Ok(ast) = graphql_ast {
                for graphql_item in ast {
                    match graphql_item {
                        GraphQLItem::Fragment(fragment) => {
                            registry.fragments.insert(fragment.name.clone(), fragment);
                        }
                        GraphQLItem::Query(query) => {
                            registry.queries.insert(query.name.clone(), query);
                        }
                    }
                }
            }
        }
    } else {
        eprintln!("Failed to parse GraphQL from file: {}", file.display());
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tree_formatter::TreeFormatter;
    use std::fs;
    use std::path::PathBuf;

    /// Formats a GraphQL registry using TreeFormatter for snapshot testing
    fn format_registry_with_tree_formatter(registry: &GraphQLRegistry) -> String {
        let mut formatter = TreeFormatter::new();

        // Git root for relative paths in snapshots
        let git_root = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .parent()
            .unwrap()
            .to_path_buf();

        formatter.add_line(0, "GraphQL Registry");

        // Format queries section
        let query_count = registry.queries.len();
        formatter.add_line(1, &format!("Queries ({})", query_count));

        if query_count > 0 {
            // Sort queries by name for consistent output
            let mut queries: Vec<_> = registry.queries.iter().collect();
            queries.sort_by(|a, b| a.key().cmp(b.key()));

            for entry in queries {
                let name = entry.key();
                let query = entry.value();
                let relative_path = query
                    .file_path
                    .strip_prefix(&git_root)
                    .unwrap_or(&query.file_path);
                formatter.add_line(2, &format!("{} ({})", name, relative_path.display()));

                // Query directives
                if !query.directives.is_empty() {
                    formatter.add_line(3, "Directives:");
                    for directive in &query.directives {
                        let emoji = match directive.directive_type {
                            crate::parsers::graphql_parser::DirectiveType::Catch => "üß§",
                            crate::parsers::graphql_parser::DirectiveType::ThrowOnFieldError => {
                                "‚ö†Ô∏è"
                            }
                        };
                        formatter.add_line(4, &format!("{:?} {}", directive.directive_type, emoji));
                    }
                }

                // Query fields
                let query_fields = query.fields();
                if !query_fields.is_empty() {
                    formatter.add_line(3, "Fields:");
                    for field in &query_fields {
                        let mut field_text = field.name.clone();
                        if !field.directives.is_empty() {
                            let directive_strs: Vec<String> = field.directives.iter().map(|d| {
                                let emoji = match d.directive_type {
                                    crate::parsers::graphql_parser::DirectiveType::Catch => "üß§",
                                    crate::parsers::graphql_parser::DirectiveType::ThrowOnFieldError => "‚ö†Ô∏è",
                                };
                                format!("{:?} {}", d.directive_type, emoji)
                            }).collect();
                            field_text.push_str(&format!(" [{}]", directive_strs.join(", ")));
                        }
                        formatter.add_line(4, &field_text);
                    }
                }

                // Query fragment spreads
                let query_fragments = query.fragments();
                if !query_fragments.is_empty() {
                    formatter.add_line(3, "Fragment Spreads:");
                    for fragment in &query_fragments {
                        let mut fragment_text = fragment.name.clone();
                        if !fragment.directives.is_empty() {
                            let directive_strs: Vec<String> = fragment.directives.iter().map(|d| {
                                let emoji = match d.directive_type {
                                    crate::parsers::graphql_parser::DirectiveType::Catch => "üß§",
                                    crate::parsers::graphql_parser::DirectiveType::ThrowOnFieldError => "‚ö†Ô∏è",
                                };
                                format!("{:?} {}", d.directive_type, emoji)
                            }).collect();
                            fragment_text.push_str(&format!(" [{}]", directive_strs.join(", ")));
                        }
                        formatter.add_line(4, &fragment_text);
                    }
                }
            }
        }

        // Format fragments section
        let fragment_count = registry.fragments.len();
        formatter.add_line(1, &format!("Fragments ({})", fragment_count));

        if fragment_count > 0 {
            // Sort fragments by name for consistent output
            let mut fragments: Vec<_> = registry.fragments.iter().collect();
            fragments.sort_by(|a, b| a.key().cmp(b.key()));

            for entry in fragments {
                let name = entry.key();
                let fragment = entry.value();
                let relative_path = fragment
                    .file_path
                    .strip_prefix(&git_root)
                    .unwrap_or(&fragment.file_path);
                formatter.add_line(2, &format!("{} ({})", name, relative_path.display()));

                // Fragment directives
                if !fragment.directives.is_empty() {
                    formatter.add_line(3, "Directives:");
                    for directive in &fragment.directives {
                        let emoji = match directive.directive_type {
                            crate::parsers::graphql_parser::DirectiveType::Catch => "üß§",
                            crate::parsers::graphql_parser::DirectiveType::ThrowOnFieldError => {
                                "‚ö†Ô∏è"
                            }
                        };
                        formatter.add_line(4, &format!("{:?} {}", directive.directive_type, emoji));
                    }
                }

                // Fragment fields
                let fragment_fields = fragment.fields();
                if !fragment_fields.is_empty() {
                    formatter.add_line(3, "Fields:");
                    for field in &fragment_fields {
                        let mut field_text = field.name.clone();
                        if !field.directives.is_empty() {
                            let directive_strs: Vec<String> = field.directives.iter().map(|d| {
                                let emoji = match d.directive_type {
                                    crate::parsers::graphql_parser::DirectiveType::Catch => "üß§",
                                    crate::parsers::graphql_parser::DirectiveType::ThrowOnFieldError => "‚ö†Ô∏è",
                                };
                                format!("{:?} {}", d.directive_type, emoji)
                            }).collect();
                            field_text.push_str(&format!(" [{}]", directive_strs.join(", ")));
                        }
                        formatter.add_line(4, &field_text);
                    }
                }

                // Fragment spreads
                let fragment_spreads = fragment.fragments();
                if !fragment_spreads.is_empty() {
                    formatter.add_line(3, "Fragment Spreads:");
                    for spread in &fragment_spreads {
                        let mut spread_text = spread.name.clone();
                        if !spread.directives.is_empty() {
                            let directive_strs: Vec<String> = spread.directives.iter().map(|d| {
                                let emoji = match d.directive_type {
                                    crate::parsers::graphql_parser::DirectiveType::Catch => "üß§",
                                    crate::parsers::graphql_parser::DirectiveType::ThrowOnFieldError => "‚ö†Ô∏è",
                                };
                                format!("{:?} {}", d.directive_type, emoji)
                            }).collect();
                            spread_text.push_str(&format!(" [{}]", directive_strs.join(", ")));
                        }
                        formatter.add_line(4, &spread_text);
                    }
                }
            }
        }

        formatter.to_string()
    }

    /// Collects all TypeScript/TSX files from a fixture directory
    fn collect_fixture_files(dir_name: &str) -> Vec<String> {
        let fixture_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .parent()
            .unwrap()
            .join("fixtures")
            .join(dir_name);

        let mut files = Vec::new();

        if let Ok(entries) = fs::read_dir(&fixture_dir) {
            let mut file_entries: Vec<_> = entries
                .filter_map(|entry| entry.ok())
                .filter(|entry| {
                    let path = entry.path();
                    path.is_file()
                        && (path.extension() == Some(std::ffi::OsStr::new("ts"))
                            || path.extension() == Some(std::ffi::OsStr::new("tsx")))
                })
                .collect();

            // Sort files by name for consistent ordering
            file_entries.sort_by_key(|entry| entry.file_name());

            for entry in file_entries {
                files.push(entry.path().to_string_lossy().to_string());
            }
        }

        files
    }

    /// Tests registry building from valid fixture files
    #[test]
    fn test_registry_from_valid_fixtures() {
        let files = collect_fixture_files("valid");
        let registry = process_files(&files);
        let formatted = format_registry_with_tree_formatter(&registry);
        insta::assert_snapshot!(formatted);
    }

    /// Tests registry building from invalid fixture files
    #[test]
    fn test_registry_from_invalid_fixtures() {
        let files = collect_fixture_files("invalid");
        let registry = process_files(&files);
        let formatted = format_registry_with_tree_formatter(&registry);
        insta::assert_snapshot!(formatted);
    }

    /// Tests registry building from edge case fixture files
    #[test]
    fn test_registry_from_edge_case_fixtures() {
        let files = collect_fixture_files("edge_cases");
        let registry = process_files(&files);
        let formatted = format_registry_with_tree_formatter(&registry);
        insta::assert_snapshot!(formatted);
    }

    /// Tests registry building from all fixture files combined
    #[test]
    fn test_registry_from_all_fixtures() {
        let mut all_files = Vec::new();
        all_files.extend(collect_fixture_files("valid"));
        all_files.extend(collect_fixture_files("invalid"));
        all_files.extend(collect_fixture_files("edge_cases"));

        let registry = process_files(&all_files);
        let formatted = format_registry_with_tree_formatter(&registry);
        insta::assert_snapshot!(formatted);
    }
}
