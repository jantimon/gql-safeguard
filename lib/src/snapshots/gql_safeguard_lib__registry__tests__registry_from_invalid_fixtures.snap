---
source: lib/src/registry.rs
expression: formatted
---
GraphQL Registry
├── Queries (4)
|   ├── GetFullUserUnprotected (fixtures/invalid/unprotected_nested.ts)
|   |   ├── Fields:
|   |   |   └── user
|   |   └── Fragment Spreads:
|   |       └── UserDetailsUnprotected
|   ├── GetUserProfileUnprotected (fixtures/invalid/missing_catch.tsx)
|   |   └── Fields:
|   |       ├── user
|   |       ├── id
|   |       ├── name
|   |       ├── avatar [ThrowOnFieldError ☄️]
|   |       └── email
|   ├── GetUserUnprotected (fixtures/invalid/unprotected_required_throw.ts)
|   |   └── Fields:
|   |       ├── user
|   |       ├── id
|   |       ├── name [RequiredThrow ☄️]
|   |       └── email
|   └── MixedQuery (fixtures/invalid/partial_protection.ts)
|       ├── Fields:
|       |   ├── user
|       |   └── id
|       └── Fragment Spreads:
|           ├── ProtectedFragment
|           └── UnprotectedFragment
└── Fragments (5)
    ├── ProtectedFragment (fixtures/invalid/partial_protection.ts)
    |   ├── Directives:
    |   |   └── Catch 🧤
    |   └── Fields:
    |       ├── sensitiveData [ThrowOnFieldError ☄️]
    |       └── otherData
    ├── UnprotectedFragment (fixtures/invalid/partial_protection.ts)
    |   └── Fields:
    |       ├── riskyField [ThrowOnFieldError ☄️]
    |       └── normalField
    ├── UserAvatarUnprotected (fixtures/invalid/unprotected_nested.ts)
    |   ├── Directives:
    |   |   └── ThrowOnFieldError ☄️
    |   └── Fields:
    |       ├── avatar
    |       └── avatarUrl
    ├── UserBasicInfoUnprotected (fixtures/invalid/unprotected_nested.ts)
    |   └── Fields:
    |       ├── id
    |       ├── name [ThrowOnFieldError ☄️]
    |       └── email
    └── UserDetailsUnprotected (fixtures/invalid/unprotected_nested.ts)
        ├── Fields:
        |   └── bio
        └── Fragment Spreads:
            ├── UserBasicInfoUnprotected
            └── UserAvatarUnprotected
